src/gsym/parser.rs

```diff
@@ -0,0 +1,424 @@
+    use super::*;
+    use std::env;
+    use std::fs::File;
+    use std::io::Read;
+    use std::mem;
+    use std::path::Path;
+    use std::ptr::copy;
+
+    #[test]
+    fn test_parse_context() {
+        let args: Vec<String> = env::args().collect();
+        let bin_name = &args[0];
+        let test_gsym = Path::new(bin_name)
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .join("data")
+            .join("test.gsym");
+        let mut gsym_fo = File::open(test_gsym).unwrap();
+        let mut data = vec![];
+
+        gsym_fo.read_to_end(&mut data).unwrap();
+        let ctx = GsymContext::parse_header(&data).unwrap();
+
+        assert_eq!(ctx.addr_at(1), 0x00000000004004b0);
+        let addrinfo = ctx.addr_info(1);
+        assert_eq!(ctx.get_str(addrinfo.name as usize), "_start");
+
+        let idx = 7;
+        assert_eq!(ctx.addr_at(idx), 0x0000000000400596);
+        let addrinfo = ctx.addr_info(idx);
+        assert_eq!(ctx.get_str(addrinfo.name as usize), "factorial");
+
+        let idx = find_address(&ctx, 0x00000000004005db);
+        assert_eq!(idx, 9);
+        let addrinfo = ctx.addr_info(idx);
+        assert_eq!(ctx.get_str(addrinfo.name as usize), "main");
+
+        let addrdata_objs = parse_address_data(addrinfo.data);
+        println!("len = {}", addrdata_objs.len());
+        for o in addrdata_objs {
+            println!("{}", o.typ);
+        }
+    }
+
+    #[test]
+    fn test_find_address() {
+        let args: Vec<String> = env::args().collect();
+        let bin_name = &args[0];
+        let test_gsym = Path::new(bin_name)
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .join("data")
+            .join("test.gsym");
+        let mut gsym_fo = File::open(test_gsym).unwrap();
+        let mut data = vec![];
+
+        const TEST_SIZE: usize = 6;
+
+        gsym_fo.read_to_end(&mut data).unwrap();
+
+        let mut addr_tab = Vec::<u8>::new();
+        addr_tab.resize(TEST_SIZE * 4, 0);
+
+        let mut ctx = GsymContext::parse_header(&data).unwrap();
+        ctx.header.num_addrs = TEST_SIZE as u32;
+        ctx.header.addr_off_size = 4;
+        ctx.header.base_address = 0;
+        // Share the same buffer.
+        ctx.addr_tab = unsafe { mem::transmute(addr_tab.as_slice()) };
+
+        let mut values: Vec<u32> = (0_u32..(TEST_SIZE as u32)).into_iter().collect();
+
+        let copy_to_addr_tab = |values: &[u32], addr_tab: &mut [u8]| {
+            let mut off = 0;
+            for v in values {
+                let dst = (&mut addr_tab[off]) as *mut u8;
+                let src = v as *const u32 as *const u8;
+                unsafe { copy(src, dst, 4) };
+                off += 4;
+            }
+        };
```

Comment:
So you want to write a couple of `u32` into a `u8`? You should be able to use `Write` in conjunction with `u32::to_le_bytes`. Please avoid unsafe code and pointer casts and arithmetic, if possible.

Reply:
Although, we support LE only. Use `u32::to_le_bytes(`) mean this test case will only run with LE platform.  So, I will change it to `to_ne_bytes()`.